utils/src/range/cover.rs

```diff
@@ -1,119 +1,119 @@
-            }
-        }
-
-        Ok(cover_subsets.into_iter())
+    ) -> Option<Vec<usize>>
+    where
+        RangeSet<T>: 'a,
+    {
+        cover(self, others).map(|sets| sets.into_iter().map(|(pos, _)| pos).collect())
     }
-}
-
-/// Error for [`RangeSetCover`].
-#[derive(Debug, thiserror::Error)]
-pub struct RangeSetCoverError {
-    kind: ErrorKind,
-    source: Option<Box<dyn std::error::Error + Send + Sync>>,
-}
 
-impl RangeSetCoverError {
-    fn new<E>(kind: ErrorKind, source: E) -> Self
+    fn cover<'a>(
+        &self,
+        others: impl IntoIterator<Item = &'a RangeSet<T>>,
+    ) -> Option<Vec<&'a RangeSet<T>>>
     where
-        E: Into<Box<dyn std::error::Error + Send + Sync>>,
+        RangeSet<T>: 'a,
     {
-        Self {
-            kind,
-            source: Some(source.into()),
-        }
+        cover(self, others).map(|sets| sets.into_iter().map(|(_, set)| set).collect())
     }
 }
 
-#[derive(Debug)]
-enum ErrorKind {
-    FailedToCover,
+struct Candidate<'a, T> {
+    /// Index in the remaining collection.
+    i: usize,
+    /// Position in the original collection.
+    pos: usize,
+    set: &'a RangeSet<T>,
+    /// The number of elements in the intersection of the set and the uncovered elements.
+    cover: usize,
 }
 
-impl fmt::Display for RangeSetCoverError {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        f.write_str("rangeset cover error: ")?;
+/// Greedy set cover algorithm.
+///
+/// Returns the fewest sets from `others` which exactly cover `query`.
+fn cover<'a, T: Copy + Ord + 'static>(
+    query: &RangeSet<T>,
+    others: impl IntoIterator<Item = &'a RangeSet<T>>,
+) -> Option<Vec<(usize, &'a RangeSet<T>)>>
+where
+    Range<T>: ExactSizeIterator<Item = T>,
+{
+    if query.is_empty() {
+        return Some(Default::default());
+    }
+
+    // Filter out rangesets that are not a subset of query.
+    let mut others: Vec<_> = others
+        .into_iter()
+        .enumerate()
+        .filter_map(|(pos, other)| {
+            if other.is_subset(query) {
+                Some((pos, other))
+            } else {
+                None
+            }
+        })
+        .collect();
 
-        match self.kind {
-            ErrorKind::FailedToCover => f.write_str("cover failure error")?,
-        }
+    if others.is_empty() {
+        return None;
+    }
 
-        if let Some(source) = &self.source {
-            write!(f, " caused by: {}", source)?;
+    let mut uncovered = query.clone();
+    let mut candidates = Vec::new();
+    let mut candidate: Option<Candidate<'_, T>> = None;
+    while !uncovered.is_empty() {
+        // Find the set with the most coverage.
+        for (i, (pos, set)) in others.iter().enumerate() {
+            let cover = set.intersection(&uncovered).len();
+            // If cover is non-empty or greater than the current candidate, update the candidate.
+            if cover > candidate.as_ref().map_or(1, |c| c.cover) {
```

Comment:
shouldn't it be `.map_or(0, |c| c.cover)`? e.g. if `uncovered` = `{[1..2], [5..6]}` and `others` contains `{[1..2]}` and `{[5..6]}`, then this will falsely return `None`?